### PR TITLE
Wizard app

### PR DESCRIPTION
**Pull request by pavle-prelic-ey**

### CODE REVIEW

File: src/components/Step3/StepThree.js [hunk 1]
```diff
@@ -1,179 +1,420 @@
...
+    dispatch({
+      type: "THIRD STEP",
+      form: stepThreeData,
+      metalworkSkills: newMetalworkSkills,
+    });
+  }, [dispatch, stepThreeData, metalworkSkills]);
+
+  const addInput = () => {
+    setCounter((counter) => {
+      return counter + 1;
```
Comment (MilanovicM) [inline src/components/Step3/StepThree.js:72]: Use implicit return here :)

File: src/components/Step3/StepThree.js [hunk 2]
```diff
@@ -1,179 +1,420 @@
...
+    );
+
+    setStepThreeData((prevState) => ({
+      ...prevState,
+      convictions: filteredConviction,
+    }));
+  };
+
+  const submitHandler = () => {
+    axios
```
Comment (MilanovicM) [inline src/components/Step3/StepThree.js:146]: Please see the previous comment about service abstraction.

File: src/components/Step3/StepThree.js [hunk 3]
```diff
@@ -1,179 +1,420 @@
...
-              <span className="ast">*</span> Do you have any agriculture skill{" "}
-            </p>
+                <textarea
+                  name="agricultureSkills"
+                  id="desc"
+                  cols="30"
+                  rows="10"
+                  onChange={handleInput}
+                  value={agricultureSkills}
+                ></textarea>
```
Comment (MilanovicM) [inline src/components/Step3/StepThree.js:201]: This can be a self closing tag.

File: src/components/Step3/StepThree.js [hunk 4]
```diff
@@ -1,179 +1,420 @@
...
               </label>
             </div>
 
-            <p className="label_s3">
-              <span className="ast">*</span> Do you have any agriculture skill{" "}
-            </p>
+            {stepThreeData.doesHaveMetalworkSkills === true ? (
+              <>
+                <InputHeading labelTitle={"What?"} />
+                <div className="input_check_wrapper">
```
Comment (MilanovicM) [inline src/components/Step3/StepThree.js:231]: These inputs here are basically all the same - except for the name and value. Maybe we could create some array with only those three values as [marketing, cutting, drilling, etc] and use map function to return input for each entry? That way we would reduce number of code lines, and if we want to add some more skills here, we don't need to write new code - just add an entry to array :)

File: src/components/Step3/StepThree.js [hunk 5]
```diff
@@ -1,179 +1,420 @@
...
-
-              <label htmlFor="joining">
-                <input type="checkbox" name="joining" id="joining" /> Joining
-              </label>
-            </div>
-
-            <p className="label_s3">
-              <span className="ast">*</span> Have you ever been convicted?{" "}
-            </p>
+            <InputHeading labelTitle={"Have you ever been convicted?"} />
```
Comment (MilanovicM) [inline src/components/Step3/StepThree.js:294]: Please see previous comments about passing text as prop.

File: src/components/ValidateSteps/ValidateStepOne.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+const ValidateStepOne = ({ firstName, lastName }) => {
+  const errors = { validated: true };
```
Comment (MilanovicM) [inline src/components/ValidateSteps/ValidateStepOne.js:2]: This file shouldn't be inside of components folder  -this is more of a helper function.

File: src/components/ValidateSteps/ValidateStepTwo.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+const ValidateStepTwo = ({ email, residencyDuration }) => {
```
Comment (MilanovicM) [inline src/components/ValidateSteps/ValidateStepTwo.js:1]: Please see comment for ValidateStepOne.js. These validation functions could maybe be one service for form validation? 
Also, the regex used could be kept inside of constants, for reusability and easier code maintenance.

File: src/components/WIzardHeader/WizardHeader.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import React from "react";
+
+const WizardHeader = ({ stepTitle }) => {
+  return (
```
Comment (MilanovicM) [inline src/components/WIzardHeader/WizardHeader.js:4]: Please use implicit return here :)

File: src/consts/statesUrl.js [hunk 9]
```diff
@@ -0,0 +1 @@
+export const stateUrl = "http://det.api.rs.ey.com/";
```
Comment (MilanovicM) [inline src/consts/statesUrl.js:1]: Good job on keeping URL in a separate file. The file name could be more generic, so we can store all potential URLs that we might add in the future.

File: src/customHooks/useFetch.js [hunk 10]
```diff
@@ -1,21 +1,24 @@
...
-  const [loading, setLoading] = useState(true);
+  const [data, setData, error, setError, loading, setLoading] =
+    useContext(NasaContext);
+
+  const getData = useCallback(async () => {
+    try {
+      const res = await axios.get(url);
+      setData(res.data);
+      res && setLoading(false);
+    } catch (error) {
```
Comment (MilanovicM) [inline src/customHooks/useFetch.js:14]: If the code in try block fails, will setLoading(false) ever get triggered? Should that be added inside of catch block?

File: src/sharedFunctions/debouncFunction.js [hunk 11]
```diff
@@ -0,0 +1,12 @@
+const debounceFunction = (func, delay) => {
```
Comment (MilanovicM) [inline src/sharedFunctions/debouncFunction.js:1]: Are we using this function elsewhere? Just asking :)

**Review by MilanovicM — CHANGES_REQUESTED** [review]
### Pavle, awesome job! The application looks great.  I really like the look and feel, and all the details - including spinner and the progress bar. Well done!
The folders are neatly organized, and kudos for using more advance concepts like - useReducer, multiple Context providers, etc...
I'm requesting changes because I wasn't able to save applicant at the end - I got 400 bad request. I suspect that you need to call different API if the applicant has no convictions. I've heard from other team members that Milos and Marija solved this, so maybe they can provide more details on this. 
As I said, that's just my assumption, so it's better if you look into that failing request again :)
Another thing that could be improved is the vertical scroll inside of carousel cards. The text inside of them would look better shortened, or maybe with "see more" or "...". 
This would improve the UX for sure ;)

Great work, again 💯

**Comment by MilanovicM** [comment]
Hi Pavle, I couldn't start the application - there is an error 
image
I think that the path to the file got broken.

File: src/App.js [hunk 12]
```diff
@@ -1,31 +1,66 @@
...
+  const readCheckHandler = () => {
+    setReadInput((readInput) => !readInput);
+  };
+  const agreeCheckHandler = () => {
+    setAgreeInput((agreeInput) => !agreeInput);
+  };
+
+  const limit = 20;
+  const key = process.env.REACT_APP_API_KEY;
+  useFetch(`${url}?api_key=${key}&count=${limit}`);
```
Comment (MilanovicM) [inline src/App.js:33]: What is the purpose of this useFetch function call? We are receiving loading, error and data from useFetch hook. Not sure if we store the values anywhere.

File: src/Helpers/ValidateSteps/ValidateStepOne.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
+const ValidateStepOne = ({ firstName, lastName }) => {
+  const errors = { validated: true };
+
+  if (!firstName.trim()) {
```
Comment (MilanovicM) [inline src/Helpers/ValidateSteps/ValidateStepOne.js:4]: These two code blocks are basically the same (4-9). Maybe we can move that to separate function, which will validate first and last name?

File: src/api/xhr.js [hunk 14]
```diff
@@ -0,0 +1,9 @@
+import axios from "axios";
+
```
Comment (MilanovicM) [inline src/api/xhr.js:2]: Good job on abstracting xhr service ;)

**Review by MilanovicM — COMMENTED** [review]
### Pavle, PR looks great :) Great job on adding tests. 

Please see my comments, there are still a few things to improve. Also, I've noticed that Address line 2 filed is marked as required, and I think it shouldn't be required. 
There are a few typos here and there, e.g. adress (d is missing), Biycicle, etc. There is Code Spell Checker extension for Visual Studio Code - maybe that can help :)

Awesome work!

**Review by MilanovicM — CHANGES_REQUESTED** [review]
### Pavle, the application looks really good. The code is also structured well, easy to read and understand!! Great job!

There are two small things that I've noticed -
1. if we choose negative residency years, and then change it to 0, it will pass the validation (the submit button will remain enabled). When that goes to BE, we get error response:
errors: {residencyDuration: ["The residency duration must be at least 1."]}
residencyDuration: ["The residency duration must be at least 1."]
message: "The given data was invalid."
2. if we have spacing in the name input - there is an error saying "Please enter text"

I think both can be easily fixed :)

**Review by MilanovicM — APPROVED** [review]
### Pavle, awesome job! 
The application looks and works great! 🥇 Bravo!!!

**Comment by pavle-prelic-ey** [comment]
> ### Pavle, awesome job!
> The application looks and works great! 🥇 Bravo!!!

Thank you!

**Pull request merged**